langchain/chat_models/base.py

```diff
@@ -178,6 +184,96 @@ async def agenerate_prompt(
             prompt_messages, stop=stop, callbacks=callbacks, **kwargs
         )
 
+    def _generate_with_cache(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
+        new_arg_supported = inspect.signature(self._generate).parameters.get(
+            "run_manager"
+        )
+        disregard_cache = self.cache is not None and not self.cache
+        if langchain.llm_cache is None or disregard_cache:
+            # This happens when langchain.cache is None, but self.cache is True
+            if self.cache is not None and self.cache:
+                raise ValueError(
+                    "Asked to cache, but no cache found at `langchain.cache`."
+                )
+            if new_arg_supported:
+                return self._generate(
+                    messages, stop=stop, run_manager=run_manager, **kwargs
+                )
+            else:
+                return self._generate(messages, stop=stop, **kwargs)
+        else:
+            params = self._get_invocation_params(stop=stop)
+            llm_string = str(sorted([(k, v) for k, v in params.items()]))
+            prompt = chat_history_as_string(messages)
+            cache_val = langchain.llm_cache.lookup(prompt, llm_string)
+            if isinstance(cache_val, list):
+                # the cache stores List[Generation]; we parse it to List[ChatGeneration]
+                cache_val_for_chat = [
+                    ChatGeneration(text=gen.text, message=AIMessage(content=gen.text))
+                    for gen in cache_val
+                ]
+                return ChatResult(generations=cache_val_for_chat)
+            else:
+                if new_arg_supported:
+                    result = self._generate(
+                        messages, stop=stop, run_manager=run_manager, **kwargs
+                    )
+                else:
+                    result = self._generate(messages, stop=stop, **kwargs)
+                langchain.llm_cache.update(prompt, llm_string, result.generations)
```

Comment:
won't we be storing ChatGenerations if _generate creates ChatGenerations? in which case do we need to do extra parsing in 218